docs: add fault injection documentation for TrafficPolicy (#710)
kgateway-dev/kgateway.dev: assets/docs/pages/resiliency/fault-injection.md

Test the resilience of your apps by injecting delays and connection failures into a percentage of your requests.

## About fault injection

> Fault injection lets you deliberately introduce failures into your request path for chaos engineering and resiliency testing. kgateway maps fault injection configuration directly to the Envoy [HTTP fault filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter), and supports the following fault types:

Reviewer: ```suggestion
Fault injection lets you deliberately introduce failures into your request path for chaos engineering and resiliency testing. {{< reuse "/docs/snippets/kgateway-capital.md" >}} maps fault injection configuration directly to the Envoy [HTTP fault filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter), and supports the following fault types:
```

* **Abort**: Terminates requests early with a specified HTTP or gRPC status code, simulating crash failures such as backend errors.
* **Delay**: Injects artificial latency before forwarding a request upstream, simulating timing failures such as network latency or overloaded backends.
* **Response rate limit**: Throttles the response body data rate to simulate slow or degraded upstream connections.

Faults are configured by creating a {{< reuse "docs/snippets/trafficpolicy.md" >}} resource and attaching it to a Gateway or HTTPRoute via `targetRefs`. When attached to a Gateway, the fault applies to all routes on that gateway. When attached to an HTTPRoute, the fault applies only to that specific route.

You can also use a route-level `disable` override to opt specific routes out of a gateway-level fault policy.

## Before you begin

{{< reuse "docs/snippets/prereq.md" >}}

## Abort {#abort}

> Use a {{< reuse "docs/snippets/trafficpolicy.md" >}} resource to abort incoming requests with a specific HTTP status code.

Reviewer: When TrafficPolicy gets imported downstream, it becomes EnterpriseTrafficPolicy, so we try to avoid phrasing that has `a` so that it doesn't render incorrectly. I will suggest rephrasing to avoid this issue throughout.

```suggestion
Abort incoming requests with a specific HTTP status code.
```

> 1. Create a {{< reuse "docs/snippets/trafficpolicy.md" >}} that aborts requests with a 503 HTTP status code. You can attach the policy to an HTTPRoute to affect a specific route, or to a Gateway to affect all routes.

Reviewer: ```suggestion
1. Configure the {{< reuse "docs/snippets/trafficpolicy.md" >}} to abort requests with a 503 HTTP status code. You can attach the policy to an HTTPRoute to affect a specific route, or to a Gateway to affect all routes.
```

       kind: HTTPRoute
       name: httpbin
     faultInjection:
       abort:
         httpStatus: 503
         percentage: 50
   EOF
   ```
   {{% /tab %}}
   {{% tab tabName="Gateway" %}}
   The following example aborts 100% of requests on the gateway. All routes on this gateway are affected.
   ```yaml
   kubectl apply -f- <<EOF
   apiVersion: {{< reuse "docs/snippets/trafficpolicy-apiversion.md" >}}
   kind: {{< reuse "docs/snippets/trafficpolicy.md" >}}
   metadata:
     name: fault-abort
     namespace: {{< reuse "docs/snippets/namespace.md" >}}
   spec:
     targetRefs:
     - group: gateway.networking.k8s.io
       kind: Gateway
       name: http
     faultInjection:
       abort:
         httpStatus: 503
         percentage: 100
   EOF
   ```
   {{% /tab %}}
   {{< /tabs >}}

   | Field | Description |
   |-------|-------------|
   | `targetRefs` | The HTTPRoute or Gateway to apply the fault injection to. |
   | `faultInjection.abort.httpStatus` | The HTTP status code to return for aborted requests. Must be between 200-599. |
   | `faultInjection.abort.percentage` | The percentage of requests to abort (0-100). Defaults to 100 if not set. |

> 2. Send a few requests to the httpbin app. Verify that some requests are rejected with a 503 HTTP response code.

Reviewer: If they did the Gateway example of 100%, wouldn't all the requests get rejected, not some?
```suggestion
2. Send a few requests to the httpbin app. Verify that requests are rejected with a 503 HTTP response code at the percentage that matches what you set in the {{< reuse "docs/snippets/trafficpolicy.md" >}}.
```

   ```
   {{% /tab %}}
   {{% tab tabName="Port-forward for local testing" %}}
   ```sh
   curl -i localhost:8080/status/200 -H "host: www.example.com"
   ```
   {{% /tab %}}
   {{< /tabs >}}

   Example output for a successful response:
   ```
   HTTP/1.1 200 OK
   access-control-allow-credentials: true
   access-control-allow-origin: *
   date: Tue, 23 Apr 2024 17:12:13 GMT
   x-envoy-upstream-service-time: 0
   server: envoy
   transfer-encoding: chunked
   ```

   Example output for an aborted request:
   ```
   HTTP/1.1 503 Service Unavailable
   content-length: 18
   content-type: text/plain
   date: Tue, 23 Apr 2024 17:12:08 GMT
   server: envoy
   ```

3. When you are done, remove the {{< reuse "docs/snippets/trafficpolicy.md" >}}.
   {{< tabs items="HTTPRoute,Gateway" tabTotal="2" >}}
   {{% tab tabName="HTTPRoute" %}}
   ```sh
   kubectl delete trafficpolicy fault-abort -n httpbin
   ```
   {{% /tab %}}
   {{% tab tabName="Gateway" %}}
   ```sh
   kubectl delete trafficpolicy fault-abort -n {{< reuse "docs/snippets/namespace.md" >}}
   ```
   {{% /tab %}}
   {{< /tabs >}}

## Delay {#delay}

> Use a {{< reuse "docs/snippets/trafficpolicy.md" >}} resource to inject latency into incoming requests.

Reviewer: ```suggestion
Delay, or inject latency into, incoming requests for a certain amount of time. 
```

> 1. Create a {{< reuse "docs/snippets/trafficpolicy.md" >}} that delays requests. You can attach the policy to an HTTPRoute to affect a specific route, or to a Gateway to affect all routes.

Reviewer: ```suggestion
1. Configure the {{< reuse "docs/snippets/trafficpolicy.md" >}} to delay requests. You can attach the policy to an HTTPRoute to affect a specific route, or to a Gateway to affect all routes.
```

   {{% tab tabName="Cloud Provider LoadBalancer" %}}
   ```sh
   curl -i http://$INGRESS_GW_ADDRESS:8080/status/200 -H "host: www.example.com:8080" -w "\nTotal time: %{time_total}s\n"
   ```
   {{% /tab %}}
   {{% tab tabName="Port-forward for local testing" %}}
   ```sh
   curl -i localhost:8080/status/200 -H "host: www.example.com" -w "\nTotal time: %{time_total}s\n"
   ```
   {{% /tab %}}
   {{< /tabs >}}

   Example output for a delayed request:
   ```
   HTTP/1.1 200 OK
   access-control-allow-credentials: true
   access-control-allow-origin: *
   date: Tue, 23 Apr 2024 17:18:51 GMT
   x-envoy-upstream-service-time: 0
   server: envoy
   transfer-encoding: chunked

   Total time: 0.512s
   ```

3. When you are done, remove the {{< reuse "docs/snippets/trafficpolicy.md" >}}.
   {{< tabs items="HTTPRoute,Gateway" tabTotal="2" >}}
   {{% tab tabName="HTTPRoute" %}}
   ```sh
   kubectl delete trafficpolicy fault-delay -n httpbin
   ```
   {{% /tab %}}
   {{% tab tabName="Gateway" %}}
   ```sh
   kubectl delete trafficpolicy fault-delay -n {{< reuse "docs/snippets/namespace.md" >}}
   ```
   {{% /tab %}}
   {{< /tabs >}}

## Response rate limit {#response-rate-limit}

> Use a {{< reuse "docs/snippets/trafficpolicy.md" >}} resource to limit the rate at which response data is sent back to the client. This is useful for simulating slow or degraded upstream connections.

Reviewer: ```suggestion
Limit the rate at which response data is sent back to the client. This way, you can simulate slow or degraded upstream connections.
```

> 1. Create a {{< reuse "docs/snippets/trafficpolicy.md" >}} that limits the response rate to 1 kbit/s for 100% of requests.

Reviewer: ```suggestion
1. Configure the {{< reuse "docs/snippets/trafficpolicy.md" >}} to limit the response rate to 1 kilobit per second for all requests.
```

       kind: HTTPRoute
       name: httpbin
     faultInjection:
       disable: {}
   EOF
   ```

3. Send a request to the httpbin app. Verify that the request succeeds with a 200 HTTP response code, even though the gateway-level policy aborts all requests.
   {{< tabs items="Cloud Provider LoadBalancer,Port-forward for local testing" tabTotal="2" >}}
   {{% tab tabName="Cloud Provider LoadBalancer" %}}
   ```sh
   curl -i http://$INGRESS_GW_ADDRESS:8080/status/200 -H "host: www.example.com:8080"
   ```
   {{% /tab %}}
   {{% tab tabName="Port-forward for local testing" %}}
   ```sh
   curl -i localhost:8080/status/200 -H "host: www.example.com"
   ```
   {{% /tab %}}
   {{< /tabs >}}

   Example output:
   ```
   HTTP/1.1 200 OK
   access-control-allow-credentials: true
   access-control-allow-origin: *
   date: Tue, 23 Apr 2024 17:12:13 GMT
   x-envoy-upstream-service-time: 0
   server: envoy
   transfer-encoding: chunked
   ```

4. When you are done, remove the resources that you created.
   ```sh
   kubectl delete trafficpolicy fault-abort-gateway -n {{< reuse "docs/snippets/namespace.md" >}}
   kubectl delete trafficpolicy fault-disable -n httpbin
   ```

> ## Cleanup

Reviewer: Since you already have delete steps in all the subtopics on this page, I would suggest removing this Cleanup section.

Author: fixed with [7131623](https://github.com/kgateway-dev/kgateway.dev/pull/710/commits/7131623c38867f74a6bf67370d18a19e7f497c2e)
